proxy/src/main/java/com/velocitypowered/proxy/bossbar/VelocityBossBar.java

```diff
@@ -0,0 +1,215 @@
+
+  private final List<Player> players;
+  private final UUID uuid;
+  private boolean visible;
+  private Component title;
+  private float progress;
+  private BossBarColor color;
+  private BossBarOverlay overlay;
+  private final VelocityBossBarManager manager;
+
+  public VelocityBossBar(
+      VelocityBossBarManager manager,
+      Component title,
+      BossBarColor color,
+      BossBarOverlay overlay,
+      float progress,
+      UUID uuid) {
+    this.title = title;
+    this.color = color;
+    this.overlay = overlay;
+    this.progress = progress;
+    if (progress > 1 || progress < 0) {
+      throw new IllegalArgumentException("Progress not between 0 and 1");
+    }
+    this.uuid = uuid;
+    visible = true;
+    players = new ArrayList<>();
+    this.manager = manager;
+  }
+
+  private VelocityBossBar() {
+    throw new UnsupportedOperationException("This class cannot be instanced.");
+  }
+
+  @Override
+  public void addPlayers(@NonNull Iterable<Player> players) {
+    players.forEach(this::addPlayer);
+  }
+
+  @Override
+  public void addPlayer(@NonNull Player player) {
+    if (!players.contains(player)) {
+      players.add(player);
+    }
+    if (player.isActive() && visible) {
+      sendPacket(player, addPacket());
+    }
+  }
+
+  @Override
+  public void removePlayer(@NonNull Player player) {
+    players.remove(player);
+    if (player.isActive()) {
+      sendPacket(player, removePacket());
+    }
+  }
+
+  @Override
+  public void removePlayers(@NonNull Iterable<Player> players) {
+    players.forEach(this::removePlayer);
+    if (players.equals(this.players)) {
+      manager.removeFromMap(uuid);
+      this.players.clear();
+    }
+  }
+
+  @Override
+  public void removeAllPlayers() {
+    removePlayers(players);
+  }
+
+  @Override
+  public @NonNull Component getTitle() {
+    return title;
+  }
+
+  @Override
+  public void setTitle(@NonNull Component title) {
+    this.title = title;
+    BossBar bar = new BossBar();
+    bar.setUuid(uuid);
+    bar.setAction(BossBar.UPDATE_NAME);
+    bar.setName(GsonComponentSerializer.INSTANCE.serialize(title));
+    players.forEach(
+        player -> {
+          if (player.isActive() && visible) {
+            sendPacket(player, bar);
+          }
+        });
+  }
+
+  @Override
+  public float getProgress() {
+    return progress;
+  }
+
+  @Override
+  public void setProgress(float progress) {
+    if (progress > 1 || progress < 0) {
+      throw new IllegalArgumentException("Progress not between 0 and 1");
+    }
+    this.progress = progress;
+    BossBar bar = new BossBar();
+    bar.setUuid(uuid);
+    bar.setAction(BossBar.UPDATE_PERCENT);
+    bar.setPercent(progress);
+    players.forEach(
+        player -> {
+          if (player.isActive() && visible) {
+            sendPacket(player, bar);
+          }
+        });
+  }
+
+  @Override
+  public @Nullable Collection<Player> getPlayers() {
+    return Collections.unmodifiableCollection(players);
```

Comment:
The practice in Velocity is always to return a immutable _copy_ of the collection in question. Try `ImmutableList.copyOf()`.